Pandas is to read and manipulte the csv files.
Numpy is for mathematical operations.

In [1]:
import pandas as pd
import numpy as np

Load the csv files

In [2]:
food = pd.read_csv("../data/raw/food.csv")
food_nutrient = pd.read_csv("../data/raw/food_nutrient.csv")
nutrient = pd.read_csv("../data/raw/nutrient.csv")
food_category = pd.read_csv("../data/raw/food_category.csv")

C:\Users\MOVVA KARTHIKA\AppData\Local\Temp\ipykernel_3188\3051990191.py:2: DtypeWarning: Columns (0: footnote) have mixed types. Specify dtype option on import or set low_memory=False.
  food_nutrient = pd.read_csv("../data/raw/food_nutrient.csv")


Check the shape(number of columns and rows). This means the csv files are loaded correctly.

In [3]:
print("Food:", food.shape)
print("Food Nutrient:", food_nutrient.shape)
print("Nutrient:", nutrient.shape)
print("Food Category:", food_category.shape)

Food: (78026, 5)
Food Nutrient: (159285, 11)
Nutrient: (477, 5)
Food Category: (28, 3)


Check the first five rows of the data to get a idea

In [4]:
food.head()

,fdc_id,data_type,description,food_category_id,publication_date
0,319874,sample_food,"HUMMUS, SABRA CLASSIC",16.0,2019-04-01
1,319875,market_acquisition,"HUMMUS, SABRA CLASSIC",16.0,2019-04-01
2,319876,market_acquisition,"HUMMUS, SABRA CLASSIC",16.0,2019-04-01
3,319877,sub_sample_food,Hummus,16.0,2019-04-01
4,319878,sub_sample_food,Hummus,16.0,2019-04-01


In [5]:
food_nutrient.head()

,id,fdc_id,nutrient_id,amount,data_points,derivation_id,min,max,median,footnote,min_year_acquired
0,2201847,319877,1051,56.30,1.0,1.0,NaN,NaN,NaN,NaN,NaN
1,2201845,319877,1002,1.28,1.0,1.0,NaN,NaN,NaN,NaN,NaN
2,2201846,319877,1004,19.00,1.0,1.0,NaN,NaN,NaN,NaN,NaN
3,2201844,319877,1007,1.98,1.0,1.0,NaN,NaN,NaN,NaN,NaN
4,2201852,319878,1091,188.00,1.0,1.0,NaN,NaN,NaN,NaN,NaN


In [6]:
nutrient.head()

,id,name,unit_name,nutrient_nbr,rank
0,2047,Energy (Atwater General Factors),KCAL,957.0,280.0
1,2048,Energy (Atwater Specific Factors),KCAL,958.0,290.0
2,1001,Solids,G,201.0,200.0
3,1002,Nitrogen,G,202.0,500.0
4,1003,Protein,G,203.0,600.0


In [7]:
food_category.head()

,id,code,description
0,1,100,Dairy and Egg Products
1,2,200,Spices and Herbs
2,3,300,Baby Foods
3,4,400,Fats and Oils
4,5,500,Poultry Products


Study the columns

In [8]:
food.columns

Index(['fdc_id', 'data_type', 'description', 'food_category_id',
       'publication_date'],
      dtype='str')

In [9]:
food_nutrient.columns

Index(['id', 'fdc_id', 'nutrient_id', 'amount', 'data_points', 'derivation_id',
       'min', 'max', 'median', 'footnote', 'min_year_acquired'],
      dtype='str')

In [10]:
nutrient.columns

Index(['id', 'name', 'unit_name', 'nutrient_nbr', 'rank'], dtype='str')

In [11]:
food_category.columns

Index(['id', 'code', 'description'], dtype='str')

Get the overview of the data in csv files with help of info()

In [12]:
food.info()

<class 'pandas.DataFrame'>
RangeIndex: 78026 entries, 0 to 78025
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   fdc_id            78026 non-null  int64  
 1   data_type         78026 non-null  str    
 2   description       78018 non-null  str    
 3   food_category_id  74216 non-null  float64
 4   publication_date  78026 non-null  str    
dtypes: float64(1), int64(1), str(3)
memory usage: 3.0 MB


In [13]:
food_nutrient.info()

<class 'pandas.DataFrame'>
RangeIndex: 159285 entries, 0 to 159284
Data columns (total 11 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   id                 159285 non-null  int64  
 1   fdc_id             159285 non-null  int64  
 2   nutrient_id        159285 non-null  int64  
 3   amount             159258 non-null  float64
 4   data_points        155714 non-null  float64
 5   derivation_id      158860 non-null  float64
 6   min                16030 non-null   float64
 7   max                16030 non-null   float64
 8   median             16632 non-null   float64
 9   footnote           22 non-null      str    
 10  min_year_acquired  30747 non-null   float64
dtypes: float64(7), int64(3), str(1)
memory usage: 13.4 MB


In [14]:
nutrient.info()

<class 'pandas.DataFrame'>
RangeIndex: 477 entries, 0 to 476
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   id            477 non-null    int64  
 1   name          477 non-null    str    
 2   unit_name     477 non-null    str    
 3   nutrient_nbr  465 non-null    float64
 4   rank          466 non-null    float64
dtypes: float64(2), int64(1), str(2)
memory usage: 18.8 KB


In [15]:
food_category.info()

<class 'pandas.DataFrame'>
RangeIndex: 28 entries, 0 to 27
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype
---  ------       --------------  -----
 0   id           28 non-null     int64
 1   code         28 non-null     int64
 2   description  28 non-null     str  
dtypes: int64(2), str(1)
memory usage: 804.0 bytes


Checking the null values count in every csv file

In [16]:
food.isnull().sum()

fdc_id                 0
data_type              0
description            8
food_category_id    3810
publication_date       0
dtype: int64

In [17]:
food_nutrient.isnull().sum()

id                        0
fdc_id                    0
nutrient_id               0
amount                   27
data_points            3571
derivation_id           425
min                  143255
max                  143255
median               142653
footnote             159263
min_year_acquired    128538
dtype: int64

In [18]:
nutrient.isnull().sum()


id               0
name             0
unit_name        0
nutrient_nbr    12
rank            11
dtype: int64

In [19]:
food_category.isnull().sum()

id             0
code           0
description    0
dtype: int64

Removing unnecessary columns before preprocessing to reduce memory usage and simplify downstream transformations. This is called Column Purning.

DataFrames(df) are created as a result of column pruning.

In [ ]:
food_df = food[                                
    [
        "fdc_id",
        "description",
        "food_category_id"
    ]
]

In [21]:
food_nutrient_df = food_nutrient[
    [
        "fdc_id",
        "nutrient_id",
        "amount"
    ]
]

In [22]:
nutrient_df = nutrient[
    [
        "id",
        "name"
    ]
]

In [23]:
food_category_df = food_category[
    [
        "id",
        "description"
    ]
]

Check the null values in new data frames.

In [24]:
food_df.isnull().sum()

fdc_id                 0
description            8
food_category_id    3810
dtype: int64

In [25]:
food_nutrient_df.isnull().sum()

fdc_id          0
nutrient_id     0
amount         27
dtype: int64

In [26]:
nutrient_df.isnull().sum()

id      0
name    0
dtype: int64

In [27]:
food_category_df.isnull().sum()

id             0
description    0
dtype: int64

the number of missing values drops dramatically because we discarded columns that aren't useful.

In food_df, description (8 null values), If a food doesn't have a name, can we recommend it? No.Without a name, it's meaningless. Decision ===>
Remove these rows.

In [28]:
food_df = food_df.dropna(subset=["description"])

In food_df, food_category_id (3810 null values), As they are id's, This needs investigation.

In [29]:
food_df[food_df["food_category_id"].isnull()].head(10)

,fdc_id,description,food_category_id
17191,748588,OLIVE OIL CONTROL COMPOSITE,NaN
17892,749289,PANERA - BACON,NaN
17897,749294,PANERA BACON TURKEY BRAVO,NaN
17898,749295,PANERA BACON TURKEY BRAVO,NaN
17899,749296,PANERA BACON TURKEY BRAVO,NaN
17900,749297,PANERA BACON TURKEY BRAVO,NaN
17901,749298,PANERA BACON TURKEY BRAVO,NaN
17902,749299,PANERA BACON TURKEY BRAVO,NaN
18998,790746,"BANANAS, UNRIPE, MEDIUM SIZE",NaN
19001,790749,"BANANAS, UNRIPE, MEDIUM SIZE",NaN


In [30]:
food_df[food_df["food_category_id"].isnull()]["description"].head(20)

17191               OLIVE OIL CONTROL COMPOSITE
17892                            PANERA - BACON
17897                 PANERA BACON TURKEY BRAVO
17898                 PANERA BACON TURKEY BRAVO
17899                 PANERA BACON TURKEY BRAVO
17900                 PANERA BACON TURKEY BRAVO
17901                 PANERA BACON TURKEY BRAVO
17902                 PANERA BACON TURKEY BRAVO
18998              BANANAS, UNRIPE, MEDIUM SIZE
19001              BANANAS, UNRIPE, MEDIUM SIZE
19002              BANANAS, UNRIPE, MEDIUM SIZE
19003              BANANAS, UNRIPE, MEDIUM SIZE
19004         BANANAS, UNRIPE, EXTRA LARGE SIZE
19005         BANANAS, UNRIPE, EXTRA LARGE SIZE
19006         BANANAS, UNRIPE, EXTRA LARGE SIZE
19014               BANANAS, RIPE, MEDIUM SIZE 
19018    BANANAS, RIPE, MEDIUM SIZE, DEHYDRATED
19019    BANANAS, RIPE, MEDIUM SIZE, DEHYDRATED
19020           BANANAS, RIPE, EXTRA LARGE SIZE
19021           BANANAS, RIPE, EXTRA LARGE SIZE
Name: description, dtype: str

In food_nutrient_df, amount (27 rows) are null values,If a nutrient amount is missing, can we use it? No. Decision ===> Drop those rows.

In [31]:
food_nutrient_df = food_nutrient_df.dropna(subset=["amount"])

We observe some duplicate rows in food_df. Let's see if they are unique or not by their id's.

In [32]:
food_df["fdc_id"].duplicated().sum()

np.int64(0)

In [33]:
food_df["description"].duplicated().sum()

np.int64(66255)

The id's are unique and descriptions alone are repeated. So, we consider the food_df rows as unique. That means they have any other differences may be.

food_category_id in food_df == id in food_category_df

This is foreign key relationship. 

There are four common joins.

Inner Join

Left Join

Right Join

Outer Join

LEFT JOIN (we will use this)

Why?

Our primary dataset is

food_df

We never want to lose foods.

LEFT JOIN

Keep ALL rows from food_df

Only add category if it exists.

In [34]:
food_category_df.head()

,id,description
0,1,Dairy and Egg Products
1,2,Spices and Herbs
2,3,Baby Foods
3,4,Fats and Oils
4,5,Poultry Products


Change the column name "description" in food_category_df. Because, we already have a column with name "description" in food_df. If, we join them directly. It creates "description_x" and "description_y" which is not clear. So, before left join. We will change the column name "description" of food_category_df.

In [35]:
food_category_df = food_category_df.rename(
    columns={
        "description": "category"
    }
)

Let's use left join

In [36]:
food_with_category = pd.merge(
    food_df,
    food_category_df,
    left_on="food_category_id",
    right_on="id",
    how="left"
)

In [37]:
food_with_category.head()

,fdc_id,description,food_category_id,id,category
0,319874,"HUMMUS, SABRA CLASSIC",16.0,16.0,Legumes and Legume Products
1,319875,"HUMMUS, SABRA CLASSIC",16.0,16.0,Legumes and Legume Products
2,319876,"HUMMUS, SABRA CLASSIC",16.0,16.0,Legumes and Legume Products
3,319877,Hummus,16.0,16.0,Legumes and Legume Products
4,319878,Hummus,16.0,16.0,Legumes and Legume Products


In [38]:
print(food_df.shape)
print(food_with_category.shape)

(78018, 3)
(78018, 5)


What happened?

Before merging:

Rows    = 78018
Columns = 3

After merging:

Rows    = 78018   Same
Columns = 5       Increased

Column number increased as we added new columns now after left join.

In [39]:
food_with_category["category"].isnull().sum()

np.int64(3810)

In [40]:
food_with_category["category"] = (
    food_with_category["category"]
    .fillna("Unknown")
)

Notice we're filling the category name, not the numeric food_category_id.

This is much cleaner.

In [41]:
food_with_category = food_with_category.drop(
    columns=[
        "food_category_id",
        "id"
    ]
)

We no longer need these columns.

food_category_id

id

Because

category

already tells us everything.

In [42]:
food_with_category.head()

,fdc_id,description,category
0,319874,"HUMMUS, SABRA CLASSIC",Legumes and Legume Products
1,319875,"HUMMUS, SABRA CLASSIC",Legumes and Legume Products
2,319876,"HUMMUS, SABRA CLASSIC",Legumes and Legume Products
3,319877,Hummus,Legumes and Legume Products
4,319878,Hummus,Legumes and Legume Products


Let's inspect the merged data in food_with_category.

How do you know your merge worked correctly?

Verified the row count remained unchanged after the left join.

Inspected the merged columns to ensure the category names were added correctly.

Checked for remaining missing values in the new category column.

Removed redundant key columns (food_category_id and id) after the merge.

In [43]:
food_with_category.head()

,fdc_id,description,category
0,319874,"HUMMUS, SABRA CLASSIC",Legumes and Legume Products
1,319875,"HUMMUS, SABRA CLASSIC",Legumes and Legume Products
2,319876,"HUMMUS, SABRA CLASSIC",Legumes and Legume Products
3,319877,Hummus,Legumes and Legume Products
4,319878,Hummus,Legumes and Legume Products


In [44]:
food_with_category.tail()

,fdc_id,description,category
78013,2751499,"Shallots, bulb, peeled, root removed, raw",Unknown
78014,2751500,"Shallots, bulb, peeled, root removed, raw",Unknown
78015,2751501,"Shallots, bulb, peeled, root removed, raw",Unknown
78016,2751502,"Shallots, bulb, peeled, root removed, raw",Unknown
78017,2751503,"Shallots, bulb, peeled, root removed, raw",Unknown


Look at food_nutrient_df.

One food appears in multiple rows.

For Apple:

Calories

Protein

Fat

Carbohydrates

Fiber

Sugar

Sodium

are stored in separate rows.

Pivoting Data

Converting long-format data into wide-format data.

In [46]:
food_nutrient_df.head(10)

,fdc_id,nutrient_id,amount
0,319877,1051,56.30
1,319877,1002,1.28
2,319877,1004,19.00
3,319877,1007,1.98
4,319878,1091,188.00
5,319878,1101,1.21
6,319878,1092,326.00
7,319878,1087,46.00
8,319878,1093,446.00
9,319878,1090,75.90


In [47]:
nutrient_df.head(20)

,id,name
0,2047,Energy (Atwater General Factors)
1,2048,Energy (Atwater Specific Factors)
2,1001,Solids
3,1002,Nitrogen
4,1003,Protein
5,1004,Total lipid (fat)
6,1005,"Carbohydrate, by difference"
7,1006,"Fiber, crude (DO NOT USE - Archived)"
8,1007,Ash
9,1008,Energy


The USDA dataset contains 477 nutrients.

Do we need all 477?

No.

For our recommendation system, we'll use these core nutrients

In [48]:
nutrient_df[
    nutrient_df["name"].str.contains("Energy", case=False, na=False)
]

,id,name
0,2047,Energy (Atwater General Factors)
1,2048,Energy (Atwater Specific Factors)
9,1008,Energy
63,1062,Energy


In [49]:
nutrient_df[
    nutrient_df["name"].str.contains("Protein", case=False, na=False)
]

,id,name
4,1003,Protein
54,1053,Adjusted Protein


In [50]:
nutrient_df[
    nutrient_df["name"].str.contains("Carbohydrate", case=False, na=False)
]

,id,name
6,1005,"Carbohydrate, by difference"
51,1050,"Carbohydrate, by summation"
73,1072,"Carbohydrate, other"
450,2039,Carbohydrates


In [51]:
nutrient_df[
    nutrient_df["name"].str.contains("Fats",case=False,na=False)
]

,id,name


In [52]:
nutrient_df[
    nutrient_df["name"].str.contains("Fat",case=False,na=False)
]

,id,name
5,1004,Total lipid (fat)
50,1049,"Solids, non-fat"
86,1085,Total fat (NLEA)
258,1257,"Fatty acids, total trans"
259,1258,"Fatty acids, total saturated"
291,1291,"Fatty acids, other than 607-615, 617-621, 624-..."
292,1292,"Fatty acids, total monounsaturated"
293,1293,"Fatty acids, total polyunsaturated"
318,1318,"Fatty acids, saturated, other"
319,1319,"Fatty acids, monounsat., other"


In [53]:
nutrient_df[
    nutrient_df["name"].str.contains("Fiber",case=False,na=False)
]

,id,name
7,1006,"Fiber, crude (DO NOT USE - Archived)"
67,1066,"Fiber, neutral detergent (DO NOT USE - Archived)"
80,1079,"Fiber, total dietary"
83,1082,"Fiber, soluble"
85,1084,"Fiber, insoluble"
444,2033,Total dietary fiber (AOAC 2011.25)
445,2034,Insoluble dietary fiber (IDF)
446,2035,Soluble dietary fiber (SDFP+SDFS)
447,2036,Soluble dietary fiber (SDFP)
448,2037,Soluble dietary fiber (SDFS)


In [54]:
nutrient_df[nutrient_df["name"].str.contains("Sugar", case=False, na=False)]

,id,name
64,1063,"Sugars, Total"
68,1067,Reducing sugars
87,1086,Total sugar alcohols
236,1235,"Sugars, added"
237,1236,"Sugars, intrinsic"
415,2000,Total Sugars


In [55]:
nutrient_df[nutrient_df["name"].str.contains("Sodium", case=False, na=False)]

,id,name
94,1093,"Sodium, Na"


In [56]:
nutrient_df[nutrient_df["name"].str.contains("Cholesterol", case=False, na=False)]

,id,name
254,1253,Cholesterol


Let's pick the target nutrients with their id's.

In [57]:
TARGET_NUTRIENTS = {
    "Energy": 1008,
    "Protein": 1003,
    "Fat": 1004,
    "Carbohydrate": 1005,
    "Fiber": 1079,
    "Sugar": 1063,
    "Sodium": 1093,
    "Cholesterol": 1253
}

Our goal is to convert this:

fdc_id	nutrient_id	amount

1102647	1008	52

1102647	1003	0.3

1102647	1004	0.2

1102647	1005	13.8

into

fdc_id	Energy	Protein	Fat	Carbohydrate	Fiber	Sugar	Sodium	Cholesterol

1102647	52	0.3	0.2	13.8	2.4	10.4	1	0

This transformation is called pivoting.

In [58]:
selected_nutrients = food_nutrient_df[
    food_nutrient_df["nutrient_id"].isin(TARGET_NUTRIENTS.values())
]

Keep only the required rows with target nutrients and delete the rest of the rows and this is called row filtering.

In [59]:
selected_nutrients.shape

(14052, 3)

In [60]:
selected_nutrients = selected_nutrients.merge(
    nutrient_df,
    left_on="nutrient_id",
    right_on="id",
    how="left"
)

In [61]:
selected_nutrients.head()

,fdc_id,nutrient_id,amount,id,name
0,319877,1004,19.0,1004,Total lipid (fat)
1,319878,1093,446.0,1093,"Sodium, Na"
2,319882,1004,18.7,1004,Total lipid (fat)
3,319884,1093,444.0,1093,"Sodium, Na"
4,319892,1004,16.6,1004,Total lipid (fat)


nutrient_id and id are duplicate columns. So, drop one of them.

In [62]:
selected_nutrients = selected_nutrients.drop(columns=["id"])

In [63]:
selected_nutrients.head()

,fdc_id,nutrient_id,amount,name
0,319877,1004,19.0,Total lipid (fat)
1,319878,1093,446.0,"Sodium, Na"
2,319882,1004,18.7,Total lipid (fat)
3,319884,1093,444.0,"Sodium, Na"
4,319892,1004,16.6,Total lipid (fat)


In [64]:
selected_nutrients["name"].unique()

<StringArray>
[          'Total lipid (fat)',                  'Sodium, Na',
        'Fiber, total dietary',                 'Cholesterol',
                     'Protein',                      'Energy',
               'Sugars, Total', 'Carbohydrate, by difference']
Length: 8, dtype: str

Rename the selected nutrients with suitable names

In [65]:
selected_nutrients["name"] = selected_nutrients["name"].replace({
    "Total lipid (fat)": "Fat",
    "Carbohydrate, by difference": "Carbohydrate",
    "Fiber, total dietary": "Fiber",
    "Sugars, Total": "Sugar",
    "Sodium, Na": "Sodium"
})

In [66]:
selected_nutrients["name"].unique()

<StringArray>
[         'Fat',       'Sodium',        'Fiber',  'Cholesterol',
      'Protein',       'Energy',        'Sugar', 'Carbohydrate']
Length: 8, dtype: str

In [67]:
selected_nutrients.head()

,fdc_id,nutrient_id,amount,name
0,319877,1004,19.0,Fat
1,319878,1093,446.0,Sodium
2,319882,1004,18.7,Fat
3,319884,1093,444.0,Sodium
4,319892,1004,16.6,Fat


In [68]:
selected_nutrients["name"].value_counts()

name
Fat             4746
Sodium          3762
Fiber           2391
Protein         1401
Cholesterol      974
Carbohydrate     376
Sugar            267
Energy           135
Name: count, dtype: int64

pivot()

Works only if there is exactly one value for each (fdc_id, nutrient) pair.

If the USDA dataset has duplicate measurements for the same nutrient, pivot() throws an error.

pivot_table()

Allows duplicates and lets us decide how to combine them.

In [69]:
nutrition_wide = selected_nutrients.pivot_table(
    index="fdc_id",
    columns="name",
    values="amount",
    aggfunc="mean"
)

index="fdc_id"

This becomes the row identifier.

Instead of

Apple
Energy

Apple
Protein

Apple
Fat

we get

Apple

once.

columns="name"

Every unique nutrient becomes a new column.

Instead of

Energy

Protein

Fat

as rows,

they become

Energy | Protein | Fat

as columns.

values="amount"

These become the actual values inside the table.

aggfunc="mean"

Suppose a food has two Energy values.

50

52

The pivot stores

51

using the average.

In [70]:
nutrition_wide.head()

name,Carbohydrate,Cholesterol,Energy,Fat,Fiber,Protein,Sodium,Sugar
fdc_id,,,,,,,,
319877,NaN,NaN,NaN,19.0,NaN,NaN,NaN,NaN
319878,NaN,NaN,NaN,NaN,NaN,NaN,446.0,NaN
319882,NaN,NaN,NaN,18.7,NaN,NaN,NaN,NaN
319884,NaN,NaN,NaN,NaN,NaN,NaN,444.0,NaN
319892,NaN,NaN,NaN,16.6,NaN,NaN,NaN,NaN


In [71]:
nutrition_wide = nutrition_wide.reset_index()

In [72]:
nutrition_wide.head()

name,fdc_id,Carbohydrate,Cholesterol,Energy,Fat,Fiber,Protein,Sodium,Sugar
0,319877,NaN,NaN,NaN,19.0,NaN,NaN,NaN,NaN
1,319878,NaN,NaN,NaN,NaN,NaN,NaN,446.0,NaN
2,319882,NaN,NaN,NaN,18.7,NaN,NaN,NaN,NaN
3,319884,NaN,NaN,NaN,NaN,NaN,NaN,444.0,NaN
4,319892,NaN,NaN,NaN,16.6,NaN,NaN,NaN,NaN


This will tell us exactly what nutrient records exist for that food.

In [73]:
selected_nutrients[
    selected_nutrients["fdc_id"] == 319877
]

,fdc_id,nutrient_id,amount,name
0,319877,1004,19.0,Fat


How Many Nutrients Does Each Food Have?

In [74]:
selected_nutrients.groupby("fdc_id")["name"].count().describe()

count    10487.000000
mean         1.339945
std          0.999098
min          1.000000
25%          1.000000
50%          1.000000
75%          1.000000
max          8.000000
Name: name, dtype: float64

In [75]:
nutrition_wide.notnull().sum()

name
fdc_id          10487
Carbohydrate      376
Cholesterol       974
Energy            135
Fat              4746
Fiber            2391
Protein          1401
Sodium           3762
Sugar             267
dtype: int64

In [76]:
selected_nutrients["name"].value_counts()

name
Fat             4746
Sodium          3762
Fiber           2391
Protein         1401
Cholesterol      974
Carbohydrate     376
Sugar            267
Energy           135
Name: count, dtype: int64

In [77]:
food_nutrient_df["nutrient_id"].nunique()

227

In [78]:
food_nutrient_df["fdc_id"].nunique()

66451

In [79]:
food["data_type"].value_counts()

data_type
sub_sample_food             65502
market_acquisition           7388
sample_food                  3890
agricultural_acquisition      810
foundation_food               436
Name: count, dtype: int64

The USDA dataset contains multiple kinds of foods.

For example

Foundation Foods

SR Legacy

Survey Foods

Experimental Foods

Branded Foods

Some food types only report a handful of nutrients.

If we are mixing every food type together, we'll naturally end up with sparse nutrition data.

We should probably keep only datasets like

Foundation

SR Legacy

because they contain comprehensive nutrient information.

These are mostly research sample records, not consumer food nutrition records.

For example:

sample_food

sub_sample_food

market_acquisition

These represent laboratory samples collected during research studies.

A single sample might only be tested for fat, another only for sodium, another only for fiber.

So, better to switch the dataset here insted of cleaning this one.